Check that all internal references are valid references to other prod…

qchaupds opened the pull request:
…ucts within the current bundle

Should check that all internal references are valid references to other products within the current bundle.
If the logical_identifier of theproduct containing the said reference does NOT belong to the bundle, the WARNING about
non-existent identifier is not raised but other WARNINGs may be raised, perhaps not being a member of any collection.

Closes NASA-PDS/validate#308

Tested in DEV for #308

**There should be no WARNING regarding a reference pointing to a logical identifier.**
% validate -R pds4.bundle  -r report_github308_bundle_valid.json   -s json -t src/test/resources/github308/valid/bundle_kaguya_derived.xml

**There should be 2 warnings for referencing a logical identifier for a product not found in this bundle for label bundle_kaguya_derived.xml**
**There should be 1 warning for referencing a logical identifier for a product not found in this bundle for label kgrs_calibrated_spectra_per1.xml**

% validate -R pds4.bundle  -r report_github308_bundle_invalid.json   -s json -t src/test/resources/github308/invalid/bundle_kaguya_derived.xml

Commit:
Check that all internal references are valid references to other products within the current bundle

1. Add test resources for github308 to src/test/resources
2. Add functions to support parsing for lid_reference, lidvid_reference and logical_identifier tags and move some constants from function to private class variables for readability in LabelUtil.java
3. Add new check if a reference is pointing to logical_identifier not in the current bundle in BundleReferentialIntegrityRule.java
4. Add debugs to CollectionReferentialIntegrityRule.java
5. Add new tests and update github51 message count in validate.feature

Refs:

https://github.com/NASA-PDS/validate/issues/308 As a user, I want to check that all Internal References are valid references to other PDS4 products within the current validating bundle

jordanpadams committed:
Add branch CI/CD action

jordanpadams committed:
fix bug in yaml

jordanpadams committed:
rename branch CI/CD

jordanpadams committed:
Simplify branch testing CICD

jordanpadams committed:
Update branch CICD to include multiple java versions

jordanpadams commented on src/main/java/gov/nasa/pds/tools/util/LabelUtil.java:
```diff
@@ -214,6 +223,84 @@ public static String getIMVersion(DOMSource source, URL context) {
+   # @param context The location of the label being parsed from.
+   * @return logicalIdentifiers A list of logical identifiers in this label.
+   */
+  public static ArrayList<String> getLogicalIdentifiers(DOMSource source, URL context) {
```
@qchaupds seems like this method is almost identical to the `getLidVidReferences` method above. can we refactor this logic?

qchaupds replied:
The 2 functions have now been refactored into a 3rd common function.
Will be reflected when code is checked in.

Noted on your comment about adding new valid test when an existing one could work already.  The only disadvantage is if a particular test resource is removed in the future, check should be made to see if other tests in the feature file is not re-using it.

jordanpadams committed:
Upgrade to setup-java@v2

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/BundleReferentialIntegrityRule.java:
```diff
@@ -88,12 +99,263 @@ public boolean isApplicable(String location) {
+      }
+  }
+
+  private boolean hasReferenceIDAndFilenameComboAdded(String singleLidorLidVidReference, URL filename) {
```
@qchaupds can we use some other data structure here instead of combining this info with a `slash`? this seems like we are trying to force this into a string, but you never know if we will need to utilize these data structures for something else in the future.

additionally, is there any way we can use the `lidvid` of the product in order to do these mappings instead of the filename? the filename will most likely work, but, the LIDVID is really the identifier for the product and avoids any possible weird collisions.

qchaupds replied:
The slash has been removed to avoid confusion.  We are trying to compare a combination of two strings so a simple concatenation was sufficient.

The mapping is now from a lid_reference or lidvid_refence to a logical_identifier (instead of from a filename).

jordanpadams reviewed (changes requested):
@qchaupds looks good. a couple changes requested.

additionally, for future issues, we need to think about our test cases a bit. i appreciate the diligence in providing a valid/invalid test case for each bug fix, but our test data is getting huge (~62M). we can start with whatever data we need to perform our testing, but if the test data already exists, or another test already accomplishes one of our test cases, we should just call that out in the PR. if it isn't obvious that another test case meets our needs, we can totally just add more data. not a huge deal. just something to keep in mind.

for instance, we already have this kaguya data elsewhere in the test suite, and we definitely already have a valid bundle run somewhere in here. we can just re-use that test data, or even those test cases, in the pull request.

jordanpadams committed:
Upgrade the pdsen-maven-parent v1.8.0

Commit:
First response to PR comments: Refactor common function and use lid_reference or lidvid_reference to map to a logical_identifier

1. Add getIdentifiersCommon() function to refactoring in LabelUtil.java
2. Use lid_reference or lidvid_reference to map to a logical_identifier instead of using a filename in BundleReferentialIntegrityRule.java
3. Remove slash when checking for combination of two string to avoid confusion in BundleReferentialIntegrityRule.java

Refs:

https://github.com/NASA-PDS/validate/issues/308 As a user, I want to check that all Internal References are valid references to other PDS4 products within the current validating bundle

Commit:
Merge branch '#308' of github.com:NASA-PDS/validate into #308

jordanpadams reviewed (approved):
looks good. thanks @qchaupds !

jordanpadams committed:
Merge branch 'master' into #308

jordanpadams committed:
Update branch-cicd.yml

jordanpadams committed:
Updating to do mvn test only
